apps/web/src/components/admin/UserGuideRow.tsx

```diff
@@ -1,28 +1,146 @@
 import dayjs from 'dayjs';
+import { useEffect, useState } from 'react';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faReply, faSpinner, faTrash } from '@fortawesome/free-solid-svg-icons';
+import { mutate } from 'swr';
 import { UserGuide } from '@ien/common';
-import { downloadGuide } from '../../services/admin';
+import { Disclosure } from '@components';
+import {
+  deleteUserGuide,
+  downloadUserGuide,
+  getUserGuideVersions,
+  restoreUserGuide,
+} from '../../services/admin';
 
 interface UserGuideProps {
   file: UserGuide;
+  showVersions: boolean;
 }
 
-export const UserGuideRow = ({ file: { name, lastModified } }: UserGuideProps) => {
-  const open = async () => {
-    const url = await downloadGuide(name);
-    window.open(url);
+export const UserGuideRow = ({ file, showVersions }: UserGuideProps) => {
+  const [versions, setVersions] = useState<UserGuide[]>();
+  const [loading, setLoading] = useState(false);
+  const [selectedVersion, setSelectedVersion] = useState<string | undefined>();
+
+  const getVersions = (open: boolean) => {
+    if (open) {
+      getUserGuideVersions(file.name)
+        .then(setVersions)
+        .catch(() => setVersions([]));
+    }
+  };
+
+  const openFile = (name: string, version?: string) => {
+    setSelectedVersion(version);
+    setLoading(true);
+    downloadUserGuide(file.name, version)
+      .then(window.open)
+      .finally(() => setLoading(false));
+  };
+
+  const getSizeWithUnit = (size: number): string => {
+    if (size > 1024 ** 2) {
+      return `${(size / 1024 ** 2).toFixed(1)}M`;
+    } else if (size > 1024) {
+      return `${(size / 1024).toFixed(1)}K`;
+    }
+    return `${size}B`;
+  };
+
+  const deleteFile = (name: string, version?: string) => {
+    setSelectedVersion(version);
+    setLoading(true);
+    deleteUserGuide(name, version)
+      .then(() => {
+        return mutate('/admin/user-guides');
+      })
+      .then(() => {
+        if (version) {
+          return getVersions(true);
+        }
+      })
+      .finally(() => setLoading(false));
+  };
+
+  const restoreFile = (name: string, version: string) => {
+    setSelectedVersion(version);
+    setLoading(true);
+    restoreUserGuide(name, version)
+      .then(() => {
+        return mutate('/admin/user-guides');
+      })
+      .then(() => getVersions(true))
+      .finally(() => setLoading(false));
+  };
+
+  useEffect(() => {
+    if (!loading) setSelectedVersion(undefined);
+  }, [loading]);
+
+  const getUserGuideRow = ({ name, lastModified, version, size }: UserGuide, index = -1) => {
+    const padding = version ? 'py-2 pl-3' : '';
+    return (
+      <div className={`flex flex-row align-middle ${padding}`} key={name + version}>
```

Comment:
are names unique? wonder if this could throw non-sense console errors complaining about the key

Reply:
I believe the name(aws object key) should be unique. If we upload one with the same name, then the old one becomes the old version of the new file.